File: db/py/nexus_db/audit.py
import hashlib
from typing import Any


def _get_val(obj: Any, key: str, default: Any = None) -> Any:
    if isinstance(obj, dict):
        return obj.get(key, default)
    return getattr(obj, key, default)


def compute_canonical_preimage(entry: Any) -> str:
    """Generate canonical pipe-delimited preimage for SHA-256 hash chaining (D-06, D-08).

    Format: prev_entry_hash|transaction_id|step_number|step_name|input_summary|output_summary|reason|is_error
    """
    is_error = _get_val(entry, "is_error", False)
    is_error_str = str(bool(is_error)).lower()

    parts = [
        str(_get_val(entry, "prev_entry_hash", "")),
        str(_get_val(entry, "transaction_id", "")),
        str(_get_val(entry, "step_number", "")),
        str(_get_val(entry, "step_name", "")),
        str(_get_val(entry, "input_summary", "")),
        str(_get_val(entry, "output_summary", "")),
        str(_get_val(entry, "reason", "")),
        is_error_str,
    ]
    return "|".join(parts)


def compute_entry_hash(entry: Any) -> str:
    """Compute SHA-256 entry_hash from canonical preimage string (D-06)."""
    preimage = compute_canonical_preimage(entry)
    return hashlib.sha256(preimage.encode("utf-8")).hexdigest()


def verify_audit_chain(entries: list[Any]) -> bool:
    """Verify cryptographic integrity and continuity of an audit trail (D-08).

    Ensures:
    1. At least 1 entry exists
    2. Step numbers are contiguous integers starting at 1
    3. Step 1 prev_entry_hash is 'GENESIS'
    4. Step N prev_entry_hash matches Step N-1 entry_hash
    5. Every entry_hash matches SHA-256 of its canonical preimage
    """
    if not entries:
        return False

    sorted_entries = sorted(entries, key=lambda x: _get_val(x, "step_number", 0))

    expected_prev = "GENESIS"
    expected_step = 1

    for entry in sorted_entries:
        step_number = _get_val(entry, "step_number")
        prev_entry_hash = _get_val(entry, "prev_entry_hash")
        entry_hash = _get_val(entry, "entry_hash")

        # 1. Verify contiguous step numbering
        if step_number != expected_step:
            return False

        # 2. Verify previous entry hash pointer
        if prev_entry_hash != expected_prev:
            return False

        # 3. Verify cryptographic SHA-256 digest
        computed_hash = compute_entry_hash(entry)
        if entry_hash != computed_hash:
            return False

        expected_prev = entry_hash
        expected_step += 1

    return True

File: db/ts/src/audit.ts
import * as crypto from 'crypto';
import { AuditEntry } from './types';

export interface AuditHashInput {
  prev_entry_hash: string;
  transaction_id: string;
  step_number: number;
  step_name: string;
  input_summary: string;
  output_summary: string;
  reason: string;
  is_error: boolean;
}

/**
 * Generate canonical pipe-delimited preimage for SHA-256 hash chaining (D-06, D-08).
 * prev_entry_hash|transaction_id|step_number|step_name|input_summary|output_summary|reason|is_error
 */
export function computeCanonicalPreimage(input: AuditHashInput): string {
  const isErrorStr = String(Boolean(input.is_error));
  return [
    input.prev_entry_hash,
    input.transaction_id,
    input.step_number.toString(),
    input.step_name,
    input.input_summary,
    input.output_summary,
    input.reason,
    isErrorStr,
  ].join('|');
}

/**
 * Compute SHA-256 entry_hash from canonical preimage string (D-06).
 */
export function computeEntryHash(input: AuditHashInput): string {
  const preimage = computeCanonicalPreimage(input);
  return crypto.createHash('sha256').update(preimage, 'utf8').digest('hex');
}

/**
 * Verify cryptographic integrity and continuity of an audit trail (D-08).
 * Ensures:
 * 1. At least 1 entry exists
 * 2. Step numbers are contiguous integers starting at 1
 * 3. Step 1 prev_entry_hash is 'GENESIS'
 * 4. Step N prev_entry_hash matches Step N-1 entry_hash
 * 5. Every entry_hash matches SHA-256 of its canonical preimage
 */
export function verifyAuditChain(entries: AuditEntry[]): boolean {
  if (!entries || entries.length === 0) {
    return false;
  }

  // Sort ascending by step_number
  const sorted = [...entries].sort((a, b) => a.step_number - b.step_number);

  let expectedPrev = 'GENESIS';
  let expectedStep = 1;

  for (const entry of sorted) {
    // 1. Verify contiguous step numbering
    if (entry.step_number !== expectedStep) {
      return false;
    }

    // 2. Verify previous entry hash
    if (entry.prev_entry_hash !== expectedPrev) {
      return false;
    }

    // 3. Verify cryptographic SHA-256 hash
    const computed = computeEntryHash(entry);
    if (entry.entry_hash !== computed) {
      return false;
    }

    expectedPrev = entry.entry_hash;
    expectedStep += 1;
  }

  return true;
}
